specs/experimental/op-contracts-manager.md

```diff
@@ -153,13 +166,98 @@ This provides the following benefits:
+  the EVM.
   - This property is contingent on the proxy and `AddressManager` code not
-  changing when OP Contracts Manager is upgraded. Both of these are not planned to
-  change.
+    changing when OP Contracts Manager is upgraded. Both of these are not planned to
+    change.
   - The OP Contracts Manager is not responsible for enforcing chain ID uniqueness, so it is acceptable
-  if this property is not preserved in future versions of the OP Contracts Manager.
+    if this property is not preserved in future versions of the OP Contracts Manager.
+
+## Upgrading
+
+### Interface
+
+#### `upgrade`
+
+The `upgrade` method is used by the Upgrade Controller to upgrade the full set of L1 contracts for
+all chains that it controls.
+
+It has the following interface:
+
+```solidity
+function upgrade(ISystemConfig[] _systemConfigs, NewChainConfig[] _newConfigs) public;
+```
+
+For each chain successfully upgraded, the following event is emitted:
+
+```solidity
+event Upgraded(uint256 indexed l2ChainId, SystemConfig indexed systemConfig);
+```
+
+This method reverts if the upgrade is not successful for any of the chains.
+
+### Implementation
+
+The high level logic of the upgrade method is as follows:
+
+1. The Upgrade Controller Safe will `DELEGATECALL` to the `OPCM.upgrade()` method.
+2. For each `_systemConfig`, the list of addresses in the chain is retrieved.
+3. For each address, a two step upgrade is used where:
+   1. the first upgrade is to an `InitializerResetter` which resets the `initialized` value.
+   1. the implementation is updated to the final address and `upgrade()` is called on that address.
+
+This approach requires that all contracts have an `upgrade()` function which sets the `initialized`
+value to `true`. The `upgrade` function body should be empty unless it is used to set a new state
+variable added to that contract since the last upgrade.
+
+#### `NewChainConfig` struct
+
+This struct is used to pass the new chain configuration to the `upgrade` method, and so it will
+vary for each release of the OP Contracts Manager, based on what (if any) new parameters are added.
+
+In practice, this struct is likely to be have a unique name for each release of the OP Contracts
+Manager.
+
+#### Requirements on the OP Chain contracts
+
+In general, all contracts used in an OP Chain SHOULD be proxied with a single shared implementation.
+This means that all values which are not constant across OP Chains SHOULD be held in storage rather
+than the bytecode of the implementation.
+
+Any contracts which do not meet this requirement will need to be deployed by the `upgrade()`
+function, increasing the cost and reducing the number of OP Chains which can be atomically upgraded.
+
+## Adding game types
+
+Because different OP Chains within a Superchain may use different dispute game types, and are
+expected to move from a permissioned to permissionless game over time, an `addGameType()` method is
```

Comment:
Is this one such example referencing this my previous question?

Reply:
Also is this another way of saying, chains in the interop set can coexist with different dispute game types?

Reply:
> Is this one such example referencing this my previous question?

Not exactly. While it is the Dispute Game contracts which contain chain specific immutable (AKA are not MCP-L1 compatible), this sentence is referring to the fact that some chains use only the `PermissionedDisputeGame`, whereas others use the `FaultDisputeGame`. In the case of chains which use the `FaultDisputeGame`, they also have the `PermissionedDisputeGame` deployed and registered so that they can fallback to the permissioned game in case there is a bug in the permissionless game (`FaultDisputeGame`).

> Also is this another way of saying, chains in the interop set can coexist with different dispute game types?

No, it's not attempting to make any statements about interop.